src/main/java/ch/njol/skript/expressions/ExprIP.java

```diff
@@ -21,61 +21,96 @@
+import org.eclipse.jdt.annotation.Nullable;
 
 @Name("IP")
-@Description("The IP address of a player.")
+@Description({"The IP address of a player, or the connected player in a <a href='events.html#connect'>connect</a>> event, " +
+		"or the pinger in a <a href='events.html#server_list_ping'>server list ping</a> event."})
 @Examples({"IP-ban the player # is equal to the next line",
 		"ban the IP-address of the player",
-		"broadcast \"Banned the IP %IP of player%\""})
+		"broadcast \"Banned the IP %IP of player%\"",
+		"",
+		"on connect:",
+		"	broadcast \"%player%: %ip%\"",
+		"",
+		"on server list ping:",
+		"	send \"%player%: %ip adress of the pinger%\" to the console"})
 @Since("1.4, 2.2-dev26 (when used in connect event)")
 public class ExprIP extends SimpleExpression<String> {
-	
+
 	static {
-		Skript.registerExpression(ExprIP.class, String.class, ExpressionType.PROPERTY, "IP[s][( |-)address[es]] of %players%",
-				"%players%'[s] IP[s][( |-)address[es]]");
+		Skript.registerExpression(ExprIP.class, String.class, ExpressionType.PROPERTY,
+				"IP[s][( |-)address[es]] of %players%",
+				"%players%'[s] IP[s][( |-)address[es]]",
+				"IP[( |-)address]",
+				"IP[( |-)address] of [the] connected player",
+				"connected [player's] IP[( |-)address]",
+				"IP[( |-)address] of [the] pinger",
+				"pinger's IP[( |-)address]");
 	}
-	
+
 	@SuppressWarnings("null")
 	private Expression<Player> players;
-	private boolean connectEvent;
-	
+
+	private boolean isConnectEvent, isServerPingEvent;
+	private int pattern;
+
 	@SuppressWarnings({"null", "unchecked"})
 	@Override
 	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
-		players = (Expression<Player>) exprs[0];
-		connectEvent = ScriptLoader.isCurrentEvent(PlayerLoginEvent.class);
+		pattern = matchedPattern;
+		isConnectEvent = ScriptLoader.isCurrentEvent(PlayerLoginEvent.class);
+		isServerPingEvent = ScriptLoader.isCurrentEvent(ServerListPingEvent.class) ||
+				(Skript.classExists("com.destroystokyo.paper.event.server.PaperServerListPingEvent") && ScriptLoader.isCurrentEvent(PaperServerListPingEvent.class));
+		if (pattern < 2) {
+			players = (Expression<Player>) exprs[0];
+		} else if (!isConnectEvent && !isServerPingEvent) {
+			Skript.error("The IP adress of pinger/connected player expression can't be used outside of a connect or server list ping event");
+			return false;
+		}
 		return true;
 	}
-	
+
 	@Override
 	@Nullable
-	protected String[] get(Event e) {
+	protected String[] get(final Event e) {
+		// If no player is specified
+		if (pattern > 1) {
+			InetAddress addr;
+			if (isConnectEvent)
+				// Return IP adress of the connected player in connect event
+				addr = ((PlayerLoginEvent) e).getAddress();
+			else
+				// Return IP adress of the pinger in server list ping event
+				addr = ((ServerListPingEvent) e).getAddress();
+			return CollectionUtils.array(addr == null ? "unknown" : addr.getHostAddress());

```

Comment:
When can it be null?

Reply:
if herobrine joins the server 😆

Reply:
Don't ask me, bensku made them like that.

Reply:
Bensku made them like that, idk.